modules/hivemq-edge-module-opcua/src/main/java/com/hivemq/edge/adapters/opcua/OpcUaProtocolAdapter.java

```diff
@@ -81,126 +91,177 @@ public OpcUaProtocolAdapter(
-                    }
+            final @NotNull ProtocolAdapterStartInput input,
+            final @NotNull ProtocolAdapterStartOutput output) {
+        if (!stopRequested.get() && startRequested.compareAndSet(false, true)) {
+            log.info("Starting OPC UA protocol adapter {}", adapterId);
+            opcUaClientConnection.start().whenComplete((ignore, throwable) -> {
+                if (throwable == null) {
+                    protocolAdapterState.setConnectionStatus(CONNECTED);
+                    output.startedSuccessfully();
+                    log.info("Successfully started OPC UA protocol adapter {}", adapterId);
+                } else {
+                    protocolAdapterState.setConnectionStatus(ERROR);
+                    startRequested.set(false);
+                    output.failStart(throwable, "Unable to connect and subscribe to the OPC UA server");
+                    log.error("Unable to connect and subscribe to the OPC UA server", throwable);
                 }
-            }
-        } else {
-            throw new IllegalStateException("Already started");
+            });
         }
     }
 
     @Override
     public void stop(final @NotNull ProtocolAdapterStopInput input, final @NotNull ProtocolAdapterStopOutput output) {
-        if (opcUaClientWrapper != null) {
-            synchronized (this) {
-                if (opcUaClientWrapper != null) {
-                    opcUaClientWrapper.stop().whenComplete((aVoid, t) -> {
-                        if (t != null) {
-                            protocolAdapterState.setErrorConnectionStatus(t, null);
-                            output.failStop(t, null);
-                        } else {
-                            opcUaClientWrapper = null;
-                            protocolAdapterState.setConnectionStatus(DISCONNECTED);
-                            output.stoppedSuccessfully();
-                        }
-                    });
+        if (startRequested.get() && stopRequested.compareAndSet(false, true)) {
+            log.info("Stopping OPC UA protocol adapter {}", adapterId);
+            opcUaClientConnection.stop(requestCounter).whenComplete((v, throwable) -> {
+                try {
+                    if (throwable == null) {
+                        protocolAdapterState.setConnectionStatus(DISCONNECTED);
+                        output.stoppedSuccessfully();
+                        log.info("Successfully stopped OPC UA protocol adapter {}", adapterId);
+                    } else {
+                        protocolAdapterState.setConnectionStatus(ERROR);
+                        output.failStop(throwable, "Unable to stop the connection to the OPC UA server");
+                        log.error("Unable to stop the connection to the OPC UA server", throwable);
+                    }
+                } finally {
+                    startRequested.set(false);
+                    stopRequested.set(false);
```

Comment:
What if the adapter is being restarted by Edge? E.g. reload config. As `startRequested.set(false)` is async, what will happen?

Reply:
The adapter looks ok, but I also checked  the ProtocolAdapterWrapper.
That one has a bug ... it's not actually waiting for the output being completed.

Reply:
Wooh, my sense is right.

Reply:
I think I am going to stop here. Fixing ProtocolAdapterWrapper will potentially drag in even more code changes, and possibly test fixes. I would rather do it as a separate ticket: https://hivemq.kanbanize.com/ctrl_board/57/cards/34176/details/